src/main/java/rx/internal/operators/OperatorOnBackpressureBuffer.java

```diff
@@ -52,122 +50,114 @@ public OperatorOnBackpressureBuffer(long capacity, Action0 onOverflow) {
-                    queue.offer(on.completed());
-                    pollQueue(wip, requested, capacity, queue, child);
-                }
+        @Override
+        public void onCompleted() {
+            if (!saturated.get()) {
+                manager.terminateAndDrain();
             }
+        }
 
-            @Override
-            public void onError(Throwable e) {
-                if (!saturated.get()) {
-                    queue.offer(on.error(e));
-                    pollQueue(wip, requested, capacity, queue, child);
-                }
+        @Override
+        public void onError(Throwable e) {
+            if (!saturated.get()) {
+                manager.terminateAndDrain(e);
             }
+        }
 
-            @Override
-            public void onNext(T t) {
-                if (!assertCapacity()) {
-                    return;
-                }
-                queue.offer(on.next(t));
-                pollQueue(wip, requested, capacity, queue, child);
+        @Override
+        public void onNext(T t) {
+            if (!assertCapacity()) {
+                return;
             }
+            queue.offer(on.next(t));
+            manager.drain();
+        }
 
-            private boolean assertCapacity() {
-                if (capacity == null) {
-                    return true;
-                }
-
-                long currCapacity;
-                do {
-                    currCapacity = capacity.get();
-                    if (currCapacity <= 0) {
-                        if (saturated.compareAndSet(false, true)) {
-                            unsubscribe();
-                            child.onError(new MissingBackpressureException(
-                                "Overflowed buffer of "
-                                + OperatorOnBackpressureBuffer.this.capacity));
-                            if (onOverflow != null) {
-                                onOverflow.call();
-                            }
-                        }
-                        return false;
-                    }
-                // ensure no other thread stole our slot, or retry
-                } while (!capacity.compareAndSet(currCapacity, currCapacity - 1));
-                return true;
+        @Override
+        public boolean accept(Object value) {
+            return on.accept(child, value);
+        }
+        @Override
+        public void complete(Throwable exception) {
+            if (exception != null) {
+                child.onError(exception);
+            } else {
+                child.onCompleted();
             }
-        };
-        
-        // if child unsubscribes it should unsubscribe the parent, but not the other way around
-        child.add(parent);
+        }
+        @Override
+        public Object peek() {
+            return queue.peek();
+        }
+        @Override
+        public Object poll() {
+            Object value = queue.poll();
+            if (capacity != null) {
```

Comment:
Sorry that I missed a case. If we poll `null`, we should not increment `capacity`.